File: libNative/src/main/java/com/munch/lib/nativelib/helper/ActivityStackHelper.java
package com.munch.lib.nativelib.helper;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import java.util.Stack;

/**
 * Created by Munch on 2018/12/16.
 */
public class ActivityStackHelper {

    private Stack<Activity> mActivityStack;

    private ActivityStackHelper() {
        mActivityStack = new Stack<>();
    }

    /**
     * registerActivityLifecycleCallbacks 此方法在Activity的super.onCreate()中调用
     */
    public void init(Application application) {
        application.registerActivityLifecycleCallbacks(new Application.ActivityLifecycleCallbacks() {
            @Override
            public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
                push(activity);
            }

            @Override
            public void onActivityStarted(Activity activity) {
            }

            @Override
            public void onActivityResumed(Activity activity) {
            }

            @Override
            public void onActivityPaused(Activity activity) {
            }

            @Override
            public void onActivityStopped(Activity activity) {
            }

            @Override
            public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
            }

            @Override
            public void onActivityDestroyed(Activity activity) {
                pop(activity);
            }
        });
    }

    /**
     * 从栈中移除Activity
     */
    private void pop(Activity activity) {
        mActivityStack.remove(activity);
    }

    /**
     * 结束Activity
     */
    public void finish(Activity activity) {
        pop(activity);
        activity.finish();
    }

    /**
     * 结束指定类名的Activity（如果存在）
     */
    public void finish(Class<? extends Activity> clazz) {
        Activity activity = getActivity(clazz);
        if (ObjectHelper.isNonEmpty(activity)) {
            finish(activity);
        }
    }

    public void push(Activity activity) {
        mActivityStack.push(activity);
    }

    /**
     * 获取栈顶的Activity
     */
    public Activity currentActivity() {
        return mActivityStack.peek();
    }

    public Activity getActivity(Class<? extends Activity> clazz) {
        for (Activity activity : mActivityStack) {
            if (activity.getClass().equals(clazz)) {
                return activity;
            }
        }
        return null;
    }

    public void finishAll() {
        for (Activity activity : mActivityStack) {
            finish(activity);
        }
    }


    public static ActivityStackHelper getInstance() {
        return Singleton.sHelper;
    }

    private static class Singleton {
        private static final ActivityStackHelper sHelper = new ActivityStackHelper();
    }
}

File: libandroid/src/main/java/com/munch/lib/android/helper/ActivityStackHelper.kt
package com.munch.lib.android.helper

import android.app.Activity
import android.app.Application
import android.os.Bundle
import androidx.lifecycle.LiveData
import com.munch.lib.android.AppHelper
import com.munch.lib.android.extend.toOrNull
import java.lang.ref.WeakReference

/**
 * 收集并提供当前Activity对象
 *
 * 是一个LiveData对象, 会在Activity进入Resume时更新为当前Activity并在Pause时更新值为null
 *
 * 对于需要当前Activity对象的方法, 当当前Activity为null时, 可在下一次更新Activity时再执行
 */
open class ActivityStackHelper<ACT : Activity> : LiveData<ACT?>() {

    private var actRef: WeakReference<ACT?>? = null
        set(value) {
            field = value
            postValue(value?.get())
        }
    protected open var isRegister = false

    override fun onActive() {
        super.onActive()
        register()
    }

    override fun onInactive() {
        super.onInactive()
        unregister()
    }

    fun register() {
        if (isRegister) return
        isRegister = true
        AppHelper.to().registerActivityLifecycleCallbacks(callback)
    }

    fun unregister() {
        if (!isRegister) return
        AppHelper.to().unregisterActivityLifecycleCallbacks(callback)
        clear()
        isRegister = false
    }

    private val callback = object : Application.ActivityLifecycleCallbacks {
        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        }

        override fun onActivityStarted(activity: Activity) {
        }

        override fun onActivityResumed(activity: Activity) {
            add(activity)
        }

        override fun onActivityPaused(activity: Activity) {
            clear()
        }

        override fun onActivityStopped(activity: Activity) {
        }

        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
        }

        override fun onActivityDestroyed(activity: Activity) {
        }
    }

    protected open fun add(act: Activity?) {
        act?.toOrNull<ACT>()?.let { actRef = WeakReference(it) }
    }

    protected open fun clear() {
        actRef?.clear()
        actRef = null
    }
}